File: client/features/movable_threads.cc
#include "movable_threads.h"

#include "client/client.h"
#include "util/debug_util.h"
#include "util/string_util.h"

#include <memory>
#include <mutex>
#include <sstream>

#include <sched.h>

namespace tetris {

MovableThreads::MovableThreads() :
    _active_mapping{nullptr}
{
    _logger = debug::Logger::get();
}

void MovableThreads::mapping_update(const MappingUpdate &mapping, const ConversionMap& /*unused*/)
{
    LOGGER->debug(" > Updating thread<->CPU assignments\n");

    /* Move the registered threads to the corresponding CPU */
    _active_mapping = std::make_unique<Mapping>(mapping);
    _available_cpus = _active_mapping->cpus;

    LOGGER->debug(" -> Using mapping %s\n", _active_mapping->name.c_str());

    {
        std::lock_guard<std::mutex> lock(_mtx);
        for (auto &t : _threads) {
            map_thread(t);
        }
    }
}

bool MovableThreads::extend_mapping(MappingsInfo &mappings)
{
    /* We don't need any extensions here, the parsed mappings should contain all necessary information */
    return false;
}

bool MovableThreads::register_thread(pid_t tid, const std::string& name)
{
    {
        std::lock_guard<std::mutex> lock(_mtx);
        /* Before adding the thread first check if it was already registered before. In this case
         * only update the naming info */
        auto it = std::find_if(_threads.begin(), _threads.end(), [tid](const ThreadInfo& t) { return t.tid == tid; });
        if (it != _threads.end()) {
            it->name = name;
            it->named = true;

            /* Sort the thread list, so that named threads are always at the beginning */
            std::sort(_threads.begin(), _threads.end(), [](const auto &t1, const auto &t2) { return t1.named; });
        } else {
            auto ti = _threads.emplace(_threads.begin(), name, tid, false, true);
            ti->managed = true;

            map_thread(*ti);
        }
    }

    LOGGER->debug("Thread registered: %d, %s\n", tid, name.c_str());


    return true;
}

bool MovableThreads::register_thread(pid_t tid)
{
    std::stringstream ss;
    ss << "thread_" << tid;

    {
        std::lock_guard<std::mutex> lock(_mtx);
        auto ti = _threads.emplace_back(ss.str(), tid, false);
        ti.managed = true;

        map_thread(ti);
    }

    LOGGER->debug("Thread registered: %d, %s\n", tid, ss.str().c_str());

    return true;
}

bool MovableThreads::unregister_thread(pid_t tid)
{
    std::lock_guard<std::mutex> lock(_mtx);

    auto it = std::find_if(std::begin(_threads), std::end(_threads),
            [tid] (const ThreadInfo& ti) { return ti.tid == tid; });

    if (it != _threads.end()) {
        _threads.erase(it);

        LOGGER->debug("Thread unregistered: %d\n", tid);

        if (it->named && it->assigned) {
            /* If this was a named thread, allow other unnamed threads to also
             * use this CPU now */
            _available_cpus.Set(it->cpu);
            for (auto &t : _threads)
                if (!t.named)
                    move_thread(t, _available_cpus);

        }

        return true;
    }

    return false;
}

void MovableThreads::map_thread(ThreadInfo &t)
{
    if (!_active_mapping)
        /* There is no mapping assigned for this application yet! */
        return;

    if (t.named) {
        LOGGER->debug(" -* Moving *named* thread %d ('%s')\n", t.tid, t.name.c_str());
        auto c = _active_mapping->cpu(t.name);

        LOGGER->debug(" --* Using CPUs: %s\n", string_util::join(c.GetList(), ",").c_str());
        if (!move_thread(t, c))
            LOGGER->warning("Failed to update thread %d\n", t.tid);

        _available_cpus^=c;
    } else {
        LOGGER->debug(" -* Moving *unnamed* thread %d ('%s')\n", t.tid, t.name.c_str());
        /* Unnamed threads can use all spare resources */
        if (_available_cpus.Size() != 0) {
            LOGGER->debug(" --* Using remaining CPUs: %s\n", string_util::join(_available_cpus.GetList(), ",").c_str());
            if (!move_thread(t, _available_cpus))
                LOGGER->warning("Failed to update thread %d\n", t.tid);
        } else {
            LOGGER->debug(" --* Using mapping CPUs: %s\n", string_util::join(_active_mapping->cpus.GetList(), ",").c_str());
            if (!move_thread(t, _active_mapping->cpus))
                LOGGER->warning("Failed to update thread %d\n", t.tid);
        }
    }
}

bool MovableThreads::move_thread(ThreadInfo &t, int cpu)
{
    cpu_set_t mask;
    CPU_ZERO(&mask);
    CPU_SET(cpu, &mask);

    if (sched_setaffinity(t.tid, sizeof(mask), &mask) != 0) {
        _logger->warning("Failed to set affinity for thread %d\n", t.tid);
        return false;
    }

    if (t.named) {
        t.cpu = cpu;
        t.assigned = true;
    }

    return true;
}

bool MovableThreads::move_thread(ThreadInfo &t, CPUThreadSet cpus)
{
    auto mask = cpus.ToCpuSetT();

    if (sched_setaffinity(t.tid, sizeof(mask), &mask) != 0) {
        _logger->warning("Failed to set affinity for thread %d\n", t.tid);
        return false;
    }

    if (t.named) {
        t.cpu = *cpus.begin();
        t.assigned = true;
    }

    return true;
}

} /* namespace tetris */

File: client_py/features/movableThreads.py
import threading
from logging import Logger
from typing import Dict

from client_py.mappingFeature import MappingFeature
from client_py.utils.cpuSets import CPUThreadSet
from client_py.utils.mapping import Mapping as MappingUpdate
from proto.tetris_pb2 import ClientMessage

# Accessing OperatingPointsInfo
MappingsInfo = ClientMessage.OperatingPointsInfo


class MovableThreads(MappingFeature):
    """
    MovableThreads class.
    """

    class ThreadInfo:
        """
        ThreadInfo class to store information about a thread.
        """

        def __init__(self, name: str, tid: int, managed: bool, named: bool = False):
            self.name = name
            self.tid = tid
            self.cpu = 0
            self.managed = managed
            self.named = named
            self.assigned = False

    def __init__(self):
        """
        Constructor for MovableThreads class.
        """
        super().__init__()
        self._logger = Logger("movable threads logger")
        self._threads = []
        self._mtx = threading.Lock()
        self._active_mapping = None
        self._available_cpus = CPUThreadSet

    def need_handshake(self) -> bool:
        """F
        Override method from MappingFeature interface.
        """
        return False

    def mapping_update(self, mapping: MappingUpdate, conv: Dict[int, int]):
        """
        Override method from MappingFeature interface.
        """
        pass  # Implement mapping_update method logic here

    def extend_mapping(self, mappings: MappingsInfo) -> bool:
        """
        Override method from MappingFeature interface.
        """
        pass  # Implement extend_mapping method logic here

    def register_thread(self, tid: int, name: str = "") -> bool:
        """
        Register a thread.

        :param tid: Thread ID.
        :param name: Thread name.
        :return: True if successful, False otherwise.
        """
        with self._mtx:
            if name:
                thread_info = self.ThreadInfo(name, tid, True, True)
            else:
                thread_info = self.ThreadInfo(str(tid), tid, True)
            self._threads.append(thread_info)
        return True  # Implement logic to return False if failed

    def unregister_thread(self, tid: int) -> bool:
        """
        Unregister a thread.

        :param tid: Thread ID.
        :return: True if successful, False otherwise.
        """
        with self._mtx:
            for thread_info in self._threads:
                if thread_info.tid == tid:
                    self._threads.remove(thread_info)
                    return True
        return False

    def map_thread(self, t: ThreadInfo):
        """
        Map a thread to a CPU.

        :param t: ThreadInfo object.
        """
        pass  # Implement map_thread method logic here

    def move_thread(self, t: ThreadInfo, cpu: int) -> bool:
        """
        Move a thread to a specific CPU.

        :param t: ThreadInfo object.
        :param cpu: CPU number.
        :return: True if successful, False otherwise.
        """
        pass  # Implement move_thread method logic here

    def move_threads(self, t: ThreadInfo, cpus: CPUThreadSet) -> bool:
        """
        Move a thread to a set of CPUs.

        :param t: ThreadInfo object.
        :param cpus: Set of CPUs.
        :return: True if successful, False otherwise.
        """
        pass  # Implement move_thread method logic here
